feat: change visibility on an existing project (#2515)

andre-code opened the pull request:
PR to include change the visibility on an existing project.

![edit visibility](https://github.com/SwissDataScienceCenter/renku-ui/assets/43388408/d6a86813-023f-47fa-baaa-5b050f087267)


![2648-4](https://github.com/SwissDataScienceCenter/renku-ui/assets/43388408/7becb8cb-3622-4c55-a0f6-0871cd07b36b)
![2648-3](https://github.com/SwissDataScienceCenter/renku-ui/assets/43388408/e0805ca9-480a-4c2f-a454-949de7a8da1f)
![2648-2](https://github.com/SwissDataScienceCenter/renku-ui/assets/43388408/88372751-7911-41ae-bc82-8cd92e161d7b)


/deploy renku-graph=development renku=auto-update/renku-ui-3.10.0 #persists

leafty commented on client/src/utils/helpers/TextUtils.ts:
```diff
@@ -0,0 +1,43 @@
+  // eslint-disable-line
+  const textValues: string[] = [];
+
+  for (const key in obj) {
```
nitpick: I would favour using `.map()` or `.flatpMap()` here:

```ts
export function extractTextFromObject(obj: Record<string, unknown>): string[] {
  const textValues = Object.values(obj).flatMap((value) => {
    if (typeof value === 'string') {
      return [value as string]
    }
    if (Array.isArray(value)) {
      const mapped = (value as unknown[]).flatMap((element) => {
        if (typeof element === 'string') {
          return [element as string]
        }
        if (typeof element === 'object') {
          return extractTextFromObject(element as Record<string, unknown>)
        }
        return []
      })
      return mapped;
    }
    if (typeof value === 'object') {
      return extractTextFromObject(value as Record<string, unknown>)
    }
    return []
  })
  return textValues;
}
```

andre-code replied:
Good idea. That would be cleaner and remove some of the complexity. I have redone this, you can find it in the last commit.

leafty commented on client/src/styles/components/_renku_general.scss:
```diff
@@ -123,3 +123,9 @@ svg:focus-visible {
     height: auto;
   }
 }
+.rk-form-section {
+  padding: 18px;
+  background-color: white;
+  border-radius: 8px;
+  margin-bottom: 20px;
+}
```
Would it be possible to add this inside a CSS or SCSS module instead? This will avoid CSS pollution (for example grouped inputs or broken because of bootstrap misconfiguration at the moment).

leafty replied:
Actually, using bootstrap classes would be the most favoured option.

andre-code replied:
I am not sure if I understood this suggestion correctly, could you please clarify which module you are referring to, one per component? In the meantime I have added the class in the css module we have created for the forms `_renku-form.scss`.

leafty reviewed (changes requested):
Overall, the logic is good.

Sorry for all the code quality comments, but these are important for readability and avoiding bugs.

andre-code commented on client/src/utils/helpers/TextUtils.test.ts:
```diff
@@ -0,0 +1,39 @@
+import { extractTextFromObject } from "./TextUtils";
+
+describe("extractTextFromObject", () => {
+  it("should extract all text values from an object with unknown keys", () => {
+    const obj = {
+      key1: "This is some text.",
+      key2: {
+        nestedKey1: "More text here.",
+        nestedKey2: {
+          deeplyNestedKey: "Even more text.",
+        },
+      },
+      key3: "Text in key 3.",
+      key4: ["Array text 1", "Array text 2"],
+      key5: {
+        nestedArray: ["Array text 3", "Array text 4"],
+        nestedObject: {
+          objectArray: ["Array text 5", "Array text 6"],
+        },
+      },
+    };
+
+    const expectedTextValues = [
+      "This is some text.",
+      "More text here.",
+      "Even more text.",
+      "Text in key 3.",
+      "Array text 1",
+      "Array text 2",
+      "Array text 3",
+      "Array text 4",
+      "Array text 5",
+      "Array text 6",
+    ];
+
+    const textValues = extractTextFromObject(obj);
+    expect(textValues).toEqual(expectedTextValues);
+  });
```
Thanks for the feedback.
As I understand the definition of expected result (expected) is defined within the Arrange section.

leafty replied:
This is a nitpick, you can ignore it if you think it makes more sense in the arrange section.

It would be nice to have:
```
<arrange>
new line
<act>
new line
<assert>
```
if possible.

andre-code commented:
> Overall, the logic is good.
> 
> Sorry for all the code quality comments, but these are important for readability and avoiding bugs.

Thanks @leafty  for your time in reviewing my PR.
I have carefully considered all your suggestions and implemented them in the latest commit, only a couple of cases are missing as they are not quite clear. I have left some comments.

leafty commented on client/src/styles/components/_renku_form.scss:
```diff
@@ -199,3 +199,10 @@
   display: flex;
   align-items: center;
 }
+
+.form-rk-section {
+  padding: 18px;
+  background-color: white;
+  border-radius: 8px;
+  margin-bottom: 20px;
+}
```
Adding global CSS classes should be avoided when possible as they leak style globally (see next comment for how to achieve the same style).

```suggestion
```

leafty commented on client/src/project/settings/ProjectSettings.present.js:
```diff
@@ -89,26 +90,42 @@ function ProjectSettingsGeneral(props) {
-            onAvatarChange={props.onAvatarChange}
-            settingsReadOnly={props.settingsReadOnly}
-          />
+          <div className="form-rk-section">
```
To avoid defining a new global CSS class, I have found that we can do:

```tsx
  return (
    <div className="form-rk-green mb-4">
      <Row className="gy-4">
        <SettingsCard>
          <ProjectTags
            tagList={props.metadata.tagList}
            onProjectTagsChange={props.onProjectTagsChange}
            settingsReadOnly={props.settingsReadOnly}
          />
        </SettingsCard>
        <SettingsCard>
          <ProjectDescription {...props} />
        </SettingsCard>
        <SettingsCard>
          <EditVisibility
            projectId={props.metadata.id}
            namespace={namespace}
            forkedProjectId={props.forkedFromProject?.id}
            visibility={props.metadata.visibility}
          />
        </SettingsCard>
        <SettingsCard>
          <ProjectAvatarEdit
            externalUrl={props.externalUrl}
            avatarUrl={props.metadata.avatarUrl}
            onAvatarChange={props.onAvatarChange}
            settingsReadOnly={props.settingsReadOnly}
            includeRequiredLabel={false}
          />
        </SettingsCard>
      </Row>
    </div>
  );
}

function SettingsCard({ children }) {
  return (
    <Col xs={12}>
      <Card>
        <CardBody className="py-2">
          <Row>
            <Col>{children}</Col>
          </Row>
        </CardBody>
      </Card>
    </Col>
  );
}
```

Note that this makes all the cards have the same spacing around them (which was not the case before).

leafty commented on client/src/components/visibility/Visibility.tsx:
```diff
@@ -132,107 +173,152 @@ const VisibilityInput = ({
   return (
     <>
-      <InputLabel text="Visibility" isRequired={isRequired} />
+      <legend>
+        {" "}
+        Visibility{" "}
+        {includeRequiredLabel ? (
+          <RequiredLabel isRequired={isRequired} />
+        ) : (
+          ""
+        )}{" "}
+      </legend>
```
The legend renders with a large font size, you can style it with:

```suggestion
      <legend className="form-label fs-6">
        Visibility{" "}
        {includeRequiredLabel && <RequiredLabel isRequired={isRequired} />}
      </legend>
```

lorenzo-cavazzi commented on client/src/components/visibility/Visibility.tsx:
```diff
@@ -132,107 +173,151 @@ const VisibilityInput = ({
 
+  const disableByNamespaceOptions = {
+    public: "",
+    private: `Public and Internal options are not available due to ${
```
```suggestion
    private: `Public and Internal not available due to ${
```
slightly shorter

lorenzo-cavazzi commented on client/src/components/visibility/Visibility.tsx:
```diff
@@ -132,107 +173,151 @@ const VisibilityInput = ({
+    private: `Public and Internal options are not available due to ${
+      isForked ? "forked project or " : ""
+    }group namespace restrictions. `,
+    internal: `Public is not available due to ${
```
```suggestion
    internal: `Public not available due to ${
```
same pattern as the previous message

lorenzo-cavazzi commented on client/src/project/new/components/Visibility.tsx:
```diff
@@ -58,4 +73,209 @@ const Visibility = ({ handlers, meta, input }: VisibilityProps) => {
+  return (
+    <Modal isOpen={isOpen} toggle={() => toggleModal()} size="lg">
+      <ModalHeader toggle={() => toggleModal()}>
+        Edit visibility to{" "}
```
```suggestion
        Change visibility to{" "}
```
Mind that this requires adjusting the e2e test -- I'll add that as a suggestion too

lorenzo-cavazzi commented on tests/cypress/e2e/project.spec.ts:
```diff
@@ -178,6 +182,39 @@ describe("display a project", () => {
+    cy.wait("@getProjectsById");
+    // should keep visibility status after cancel update
+    cy.get_cy("visibility-private").click();
+    cy.get(".modal-title").should("contain.text", "Edit visibility to Private");
```
```suggestion
    cy.get(".modal-title").should("contain.text", "Change visibility to Private");
```

lorenzo-cavazzi commented on client/src/project/new/components/Visibility.tsx:
```diff
@@ -58,4 +73,209 @@ const Visibility = ({ handlers, meta, input }: VisibilityProps) => {
+      <LoadingLabel text="Updating visibility" />
+    ) : (
+      <>
+        Please note that users will not need your explicit permission to see
+        this project. If you need more information about visibility, please
```
This message is correct for public and internal projects but misleading for private projects; it's actually the opposite.
In that case, I would change it to something like "Users will not find this project anymore unless you explicitly add them to the Members list", with a link to the members page in GitLab

![image](https://github.com/SwissDataScienceCenter/renku-ui/assets/43481553/12ee058b-058b-4e37-ac51-607a3698580e)

lorenzo-cavazzi commented on client/src/project/new/components/Visibility.tsx:
```diff
@@ -58,4 +73,209 @@ const Visibility = ({ handlers, meta, input }: VisibilityProps) => {
+  message,
+  visibility,
+}: EditVisibilityModalConfirmationProps) {
+  const buttons = !isError && !isLoading && !isSuccess && (
```
Should we show a "Close" button in all the other cases?
Feel free to disagree and keep the current solution here :slightly_smiling_face:

lorenzo-cavazzi reviewed (changes requested):
Very useful addition!
I added just a couple of inline comments about working or tiny changes.

There is just one additional thing I would change. We should ignore clicks to the current visibility. E.G. if the project is Private already and the user tries to click on Private, we should not open the modal to confirm changing it. Mind that this happens only when clicking on the label, while the radio button already ignores the click correctly (see video)

![Peek 2023-07-20 19-04](https://github.com/SwissDataScienceCenter/renku-ui/assets/43481553/22de14ef-de97-4ed2-bc0f-27a4c65e9b08)

andre-code commented:
> Very useful addition! I added just a couple of inline comments about working or tiny changes.
> 
> There is just one additional thing I would change. We should ignore clicks to the current visibility. E.G. if the project is Private already and the user tries to click on Private, we should not open the modal to confirm changing it. Mind that this happens only when clicking on the label, while the radio button already ignores the click correctly (see video)
> 
> ![Peek 2023-07-20 19-04](https://user-images.githubusercontent.com/43481553/254959493-22de14ef-de97-4ed2-bc0f-27a4c65e9b08.gif)

Thanks @lorenzo-cavazzi for your suggestions, I have implemented them and also those discussed in the product ui meeting.
In addition I have included here the endpoint kg to update the visibility, if the project is not yet activated we show an Alert.

lorenzo-cavazzi commented on client/src/project/new/components/Visibility.tsx:
```diff
@@ -58,4 +75,231 @@ const Visibility = ({ handlers, meta, input }: VisibilityProps) => {
+
+  if (!isLoadingProject && !isFetchingProject && !projectData)
+    return (
+      <WarnAlert>
+        Knowledge Graph integration must be enabled to change visibility from
+        the RenkuLab web interface.
+      </WarnAlert>
```
Should we make this non-dismissible? Otherwise, we are left with a white space.

![Peek 2023-07-24 14-39](https://github.com/SwissDataScienceCenter/renku-ui/assets/43481553/5218847e-18da-4c4b-b1dc-6d8a90a44e9a)

lorenzo-cavazzi reviewed (changes requested):
This looks good! It does include #2675 already, right?

I added a couple of minor comments inline. We can then merge it

leafty commented on client/src/components/formlabels/FormLabels.tsx:
```diff
@@ -69,6 +75,7 @@ const HelperLabel = ({ text }: LabelProps) => {
 };
 
 const InputHintLabel = ({ text }: LabelProps) => {
+  // TODO: remove this component
```
See https://github.com/SwissDataScienceCenter/renku-ui/pull/2648#discussion_r1267722707. This adds one level of indirection for no benefit. Just use `FormText`.

leafty commented on client/src/components/formlabels/FormLabels.tsx:
```diff
@@ -68,8 +68,12 @@ const HelperLabel = ({ text }: LabelProps) => {
+    <FormText className="input-hint">
+      {text} {children}
+    </FormText>
+  );
```
Done in 84a42dde5fcb849d5ce683668a7632c2fc20c111

leafty commented on client/src/components/visibility/Visibility.tsx:
```diff
@@ -132,107 +173,149 @@ const VisibilityInput = ({
+        </div>
+      </div>
+      {isDisabled ? (
+        <ThrottledTooltip
```
Use `UncontrolledTooltip` from `reactstrap`.

lorenzo-cavazzi reviewed (approved):
Lgtm! :rocket:
